Majors crud operations

rohanmaheshwari430 opened the pull request:
### Summary <!-- Required -->

This PR builds off of the firebase operations done in the "Initialized Firebase and Defined API Endpoints" PR. This time, I implemented the Majors schema and CRUD Operations for the Majors collection. I also created the majors collection in firebase and populated it with with a couple of test objects to ensure API functionality. I also reorganized the types defined in the server file to a new file to ensure clean code. All endpoints were tested via postman. The object structure and request parameters can be found in in the attached Postman collection. 

### Test Plan <!-- Required -->

To test these endpoints, cd into the server directory and type the command "yarn start". It may return an error saying you are missing the service account file, in which case please reach out to Jessica to get instructions on receiving a service account file. Make sure to rename it to the service account file name that the error mentions. If no error was returned, copy this Postman link at https://www.getpostman.com/collections/c5ea29567bc40885aa52 and import it as a collection in your personal Postman workspace. Now, you should be all set to test the endpoints. Each of these requests targets one of the four Majors-related API endpoints. Feel free to change the request body or query parameters to test different data.

https://www.getpostman.com/collections/c5ea29567bc40885aa52

GET Majors
<img width="754" alt="Screen Shot 2021-04-09 at 12 57 19 PM" src="https://user-images.githubusercontent.com/46412997/114215044-1fc02500-9933-11eb-8442-d02356034713.png">

DELETE Major
<img width="866" alt="Screen Shot 2021-04-09 at 12 57 48 PM" src="https://user-images.githubusercontent.com/46412997/114215105-31093180-9933-11eb-9a21-ec0d6708715d.png">

CREATE Major
<img width="868" alt="Screen Shot 2021-04-09 at 12 58 15 PM" src="https://user-images.githubusercontent.com/46412997/114215162-40887a80-9933-11eb-9800-d4bf97cd4de2.png">

UPDATE Major
<img width="869" alt="Screen Shot 2021-04-09 at 12 58 38 PM" src="https://user-images.githubusercontent.com/46412997/114215195-4ed69680-9933-11eb-9495-82a2a68bc825.png">



### Notes <!-- Optional -->

- It may be worth creating separate files for each collection's API endpoints.

rohanmaheshwari430 committed:
wrote CRUD ops for Majors

Wrote all CRUD Ops for majors, implemented majors schema as type, and organized all types into one file.

rohanmaheshwari430 committed:
Wrote script for importing courses to firestore

Script.ts contains a import courses function and a csv reader to take courses from a csv file (in the website_data_csv folder), parse the information into a list of JSON objects, and create a doc for each object in the courses collection in firestore. Also created a command "yarn script" in the package.json to run this script.

rohanmaheshwari430 committed:
Bug fixes on courses script

Fixed issue where comma was appearing in design areas.

rohanmaheshwari430 committed:
implemented clubs endpoints and schema

Wrote CRUD operations for clubs and created a type for a club in types.ts

rohanmaheshwari430 committed:
renamed script file for courses

renamed script to import courses to reduce confusion on import script for courses, majors, and club.

rohanmaheshwari430 committed:
renamed script.ts

rohanmaheshwari430 committed:
added import script for clubs and majors/minors

created two files for the script to import clubs and majors/minors into their respective collections. The csv files are not yet complete, so the import has not occurred.

zachary-kent reviewed (commented):
Not positive about this, but I'm pretty sure most checks for example "field == null" should be "field === undefined" I'm pretty sure that if a field is not present in an object, it will be undefined rather than null. Also, you should use === as opposed to == for stricter equality checking and to avoid undesired type coercion.

rohanmaheshwari430 committed:
Merge branch 'master' into Majors-CRUD-Operations

rohanmaheshwari430 committed:
Merge pull request #23 from cornell-dti/import-courses-script

Wrote script for importing courses to firestore

rohanmaheshwari430 committed:
Merge branch 'Majors-CRUD-Operations' into Clubs-CRUD-Operations

rohanmaheshwari430 committed:
Merge pull request #24 from cornell-dti/Clubs-CRUD-Operations

Clubs crud operations

rohanmaheshwari430 committed:
Merge branch 'Majors-CRUD-Operations' into majors-clubs-script

rohanmaheshwari430 committed:
Merge pull request #26 from cornell-dti/majors-clubs-script

Wrote Majors and clubs script

rohanmaheshwari430 committed:
Update server.ts

rohanmaheshwari430 committed:
wrote CRUD endpoints for Events

rohanmaheshwari430 commented:
> Not positive about this, but I'm pretty sure most checks for example "field == null" should be "field === undefined" I'm pretty sure that if a field is not present in an object, it will be undefined rather than null. Also, you should use === as opposed to == for stricter equality checking and to avoid undesired type coercion.

Thanks for the critic! So I went back and tested my endpoint by purposely not including one of the fields, and I did end up getting a failure response saying that one or more fields was missing. So it seems like 'field == null' does the trick? I think this may be because "undefined" still means there is some value for that field, it is just not comprehendable. Null means there is no value for that field. What do you think?

rohanmaheshwari430 committed:
wrote events script

rohanmaheshwari430 committed:
Update package.json

rohanmaheshwari430 committed:
snake case to camel case

cleaned code to satisfy eslint requirements

rohanmaheshwari430 committed:
Merge pull request #28 from cornell-dti/events-import-script

Events import script

rohanmaheshwari430 committed:
more snake case to camel case

cleaned code to satisfy eslint requirements

rohanmaheshwari430 committed:
Update courses.csv

rohanmaheshwari430 committed:
bug fixes with scripts

rohanmaheshwari430 committed:
Merge pull request #27 from cornell-dti/events-crud-operations

Events crud operations

rohanmaheshwari430 committed:
bug fixes on get course

course id was not being incremented

rohanmaheshwari430 committed:
added content nested object to getCourses

when creating a course, we require the user to provide a nested content object, however, when retrieving courses via /getCourse, we did not return a content object. thus, we added a content nested object in this commit.

rohanmaheshwari430 committed:
Merge branch 'master' into Majors-CRUD-Operations

katrina433 committed:
snake case to camel case

rohanmaheshwari430 committed:
Update types.ts

rohanmaheshwari430 committed:
Merge branch 'Majors-CRUD-Operations' of https://github.com/cornell-dti/DesignAtCornell into Majors-CRUD-Operations

rohanmaheshwari430 committed:
Update server.ts

katrina433 committed:
Update CourseBubble.tsx

meganyin13 commented:
@rohanmaheshwari430 great work on this PR! But I will back up Zak's point that you should use `===` _always_. This is just one of the ugly parts of JavaScript/TypeScript that ES6 tries to fix by adding `===`. Triple equals does strict equality checking whereas double equals is just checking if something is "truthy" or "falsy". This means JS will try to coerce whatever you're comparing to a true/false value and that means things like 0 or "" (empty string) can end up as false. Currently you're seeing `== undefined` works fine because it is falsy but it's better to strictify and explicitly say you want it to be `=== undefined` since things like `0 == undefined` or `"" == undefined` due to the falsy comparison. 
You can read more about the bad parts of JS here https://dev.to/khalalw/javascript-quirks-explaining-the-bad-parts-299j and this is the video everyone refers to when saying JS kinda sucks haha https://www.destroyallsoftware.com/talks/wat

Still great work on this PR and I'm happy to see such careful reviews!😁

rohanmaheshwari430 committed:
equality update

chnaged == null to === null for stricter type checking.

rohanmaheshwari430 commented:
> @rohanmaheshwari430 great work on this PR! But I will back up Zak's point that you should use `===` _always_. This is just one of the ugly parts of JavaScript/TypeScript that ES6 tries to fix by adding `===`. Triple equals does strict equality checking whereas double equals is just checking if something is "truthy" or "falsy". This means JS will try to coerce whatever you're comparing to a true/false value and that means things like 0 or "" (empty string) can end up as false. Currently you're seeing `== undefined` works fine because it is falsy but it's better to strictify and explicitly say you want it to be `=== undefined` since things like `0 == undefined` or `"" == undefined` due to the falsy comparison.
> You can read more about the bad parts of JS here https://dev.to/khalalw/javascript-quirks-explaining-the-bad-parts-299j and this is the video everyone refers to when saying JS kinda sucks haha https://www.destroyallsoftware.com/talks/wat
> 
> Still great work on this PR and I'm happy to see such careful reviews!😁

Thank you for this advice @meganyin13 . Sorry for the misunderstanding (@zachary0kent). I will make the changes accordingly to enforce this stricter typing.

rohanmaheshwari430 committed:
type correction

rohanmaheshwari430 committed:
type correction (clubs and events)

rohanmaheshwari430 committed:
removed redudant type

rohanmaheshwari430 committed:
Merge branch 'master' into Majors-CRUD-Operations

rohanmaheshwari430 committed:
Update package.json

rohanmaheshwari430 committed:
syntax error

rohanmaheshwari430 committed:
formatter

rohanmaheshwari430 committed:
lint fixes

zachary-kent reviewed (approved):
Thanks for changing the '==' to '==='! This PR looks awesome and all of your endpoints seem to be implemented perfectly--I used most of them while writing code for other branches I was working on and ran into no issues. I also appreciate that you changed all of the snake case to camel case--I know that must've been tedious. My only suggestion is that you should probably avoid indexed for loops, replacing them with functional calls to map() and filter(), though it's probably not worth it to change it if it works now. Also, it might be helpful to document the type of the object/array each endpoint sends back. Once again, great job!

jessfeng commented on server/server.ts:
```diff
@@ -133,6 +161,300 @@ app.post('/updateCourse', async (req, res) => {
+    res.send('One or more fields is missing.');
+  } else {
+    majors.doc(title).update({ field: content });
+    res.send({ success: true, data: [] });
```
If data is just going to be an empty array, it might be better to just not include it when sending a response back.

rohanmaheshwari430 replied:
Thanks for the comment, I agree with that change and have implemented it!

jessfeng reviewed (commented):
Looks great! Just had one comment about the response for the update endpoint, but other than that, we should be good to go!

rohanmaheshwari430 committed:
removed empty data array for update endpoints

rohanmaheshwari430 committed:
formatting
